sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/basicLogicalOperators.scala

```diff
@@ -655,12 +666,76 @@ case class Aggregate(
     }
   }
 
+  private def expandGroupingReferences(e: Expression): Expression = {
+    e match {
+      case _: AggregateExpression => e
+      case _ if PythonUDF.isGroupedAggPandasUDF(e) => e
+      case g: GroupingExprRef => groupingExpressions(g.ordinal)
+      case _ => e.mapChildren(expandGroupingReferences)
+    }
+  }
+
+  lazy val aggregateExpressionsWithoutGroupingRefs = {
+    aggregateExpressions.map(expandGroupingReferences(_).asInstanceOf[NamedExpression])
+  }
+
   override lazy val validConstraints: ExpressionSet = {
     val nonAgg = aggregateExpressions.filter(_.find(_.isInstanceOf[AggregateExpression]).isEmpty)
     getAllValidConstraints(nonAgg)
   }
 }
 
+object Aggregate {
+  private def collectComplexGroupingExpressions(groupingExpressions: Seq[Expression]) = {
+    groupingExpressions.zipWithIndex
+      .foldLeft(mutable.Map.empty[Expression, (Expression, Int)]) {
+        case (m, (ge, i)) =>
+          if (ge.deterministic && !ge.foldable && ge.children.nonEmpty &&
```

Comment:
IIRC the analyzer enforces that `groupingExpressions` are deterministic?

And why do we exclude `Attribute` to be `GroupingExprRef`?

Reply:
Indeed, `PullOutNondeterministic` is taking care of them.

`GroupingExprRef` is not needed when it would refer to a simple `Attribute`. The purpose of `GroupingExprRef`s to wrap expressions that could change during optimization, I don't think that a simple attribute could change.
Also, `GroupingExprRef` is not a `NamedExpression` now. If we wanted to allow `GroupingExprRef` to refer to a simple attribute in aggregate expressions then either it should implement `NamedExpression` or it should be wrapped into a new `Alias`.